### PR TITLE
Add AssetId.resolve

### PR DESCRIPTION
**Pull request by matanlurey**
Closes https://github.com/dart-lang/build/issues/192.

### CODE REVIEW

File: build/lib/src/asset/id.dart [hunk 1]
```diff
@@ -22,13 +22,57 @@ class AssetId implements Comparable<AssetId> {
...
   /// where possible.
   AssetId(this.package, String path) : path = _normalizePath(path);
 
+  /// Creates a new [AssetId] from an [uri].
+  ///
+  /// This gracefully handles `package:` URIs, which is useful when creating an
+  /// [AssetId] from an `import` or `export` directive pointing to a package:
+  /// ```dart
+  /// AssetId assetOfDirective(UriReferencedElement element) {
+  ///   return new AssetId.resolve(element.uri);
```
Comment (jakemac53) [inline build/lib/src/asset/id.dart]: Could we pass a sensible from here using the containing library?

File: build/lib/src/asset/id.dart [hunk 2]
```diff
@@ -22,13 +22,57 @@ class AssetId implements Comparable<AssetId> {
...
+  /// }
+  /// ```
+  factory AssetId.resolve(String uri, {AssetId from}) {
+    final parsedUri = Uri.parse(uri);
+    // If a dart:, package:, or file: URI was used, assume an absolute path.
+    if (parsedUri.hasScheme) {
+      if (parsedUri.scheme == 'package') {
+        return new AssetId(
+          parsedUri.pathSegments.first,
+          pathos.joinAll(parsedUri.pathSegments.skip(1)),
```
Comment (jakemac53) [inline build/lib/src/asset/id.dart]: Nit: we don't use trailing comments in this package for constructors. It's really unfortunate that this "feature" introduced effectively multiple styles 😢.

File: build/lib/src/asset/id.dart [hunk 3]
```diff
@@ -22,13 +22,57 @@ class AssetId implements Comparable<AssetId> {
...
+  /// Creates a new [AssetId] from an [uri].
+  ///
+  /// This gracefully handles `package:` URIs, which is useful when creating an
+  /// [AssetId] from an `import` or `export` directive pointing to a package:
+  /// ```dart
+  /// AssetId assetOfDirective(UriReferencedElement element) {
+  ///   return new AssetId.resolve(element.uri);
+  /// }
+  /// ```
+  ///
```
Comment (kevmoo) [inline build/lib/src/asset/id.dart]: Extra new-line?

File: build/lib/src/asset/id.dart (hunk 2 above)
Comment (matanlurey) [inline build/lib/src/asset/id.dart]: No worries, sorry, bad habit from Angular where we have a crazy amount of args.

File: build/lib/src/asset/id.dart (hunk 1 above)
Comment (matanlurey) [inline build/lib/src/asset/id.dart]: `element.librarySource.source.uri.toString()`?

File: build/lib/src/asset/id.dart (hunk 3 above)
Comment (matanlurey) [inline build/lib/src/asset/id.dart]: Done

File: build/lib/src/asset/id.dart (hunk 1 above)
Comment (jakemac53) [inline build/lib/src/asset/id.dart]: So `element.librarySource.source` used to be an `AssetBasedSource`, which had an `id` field. However I think with the summaries resolver that may not be the case which makes things a bit more complicated. Would you just file an issue to support more than `package:` URIs and we can follow up.

Comment (matanlurey) [inline build/lib/src/asset/id.dart]: Sure. Here, right?

Comment (matanlurey) [inline build/lib/src/asset/id.dart]: https://github.com/dart-lang/build/issues/208

**Comment by matanlurey** [comment]
@natebosch Anything else? Ok to squash and merge?

File: build/lib/src/asset/id.dart [hunk 4]
```diff
@@ -22,13 +22,49 @@ class AssetId implements Comparable<AssetId> {
...
+  /// Optionally, specifying the origin asset ([from]) will also allow resolving
+  /// relative to another asset - otherwise an [ArgumentError] will be thrown.
+  /// ```dart
+  /// AssetId assetOfDirective(AssetId origin, UriReferencedElement element) {
+  ///   return new AssetId.resolve(element.uri, from: origin);
+  /// }
+  /// ```
+  factory AssetId.resolve(String uri, {AssetId from}) {
+    final parsedUri = Uri.parse(uri);
+    // If a dart:, package:, or file: URI was used, assume an absolute path.
```
Comment (natebosch) [inline build/lib/src/asset/id.dart]: [nit] this comment doesn't look like it matches behavior. Maybe remove it?

Comment (matanlurey) [inline build/lib/src/asset/id.dart]: Done.

**Comment by matanlurey** [comment]
Thanks! This will help :)

**Pull request merged**